Low stock threshold for variations

peterfabian opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #22706.
Closes #27371.

This PR adds support for low stock threshold for variations.

It needs a bit of documentation to be added to our public user-facing docs.

In principle, I tried to keep the changes to the minimum. 
I've added a field to the variation admin UI to allow setting up the low stock amount iff the variation is managing its stock:
![Screenshot 2021-03-11 at 16 25 06](https://user-images.githubusercontent.com/2207451/110810985-96abc480-8286-11eb-99a0-f02ef7a8a852.png)

UPDATE: 

If the value is not set, it falls back to the parent product's setting (if it's set to manage stock) and only then to store-wide setting (you might notice that the number is a slightly lighter shade than other inputs):
![Screenshot 2021-03-11 at 16 27 31](https://user-images.githubusercontent.com/2207451/110811216-c8bd2680-8286-11eb-963d-77738d6f55fa.png)

The setting comes from WC > Product > Inventory settings:
![Screenshot 2021-03-11 at 16 29 32](https://user-images.githubusercontent.com/2207451/110811465-00c46980-8287-11eb-8c08-3137e8ecda68.png)

If the low stock amount is set both on the parent variable product (manage stock is on, low stock amount set to X) and on the variation (also manage stock is on, low stock amount set to Y), the amount from the variation (Y) gets used. If there is no value set on the variation, the amount X from the parent product gets used.

I also enabled setting the **variation low stock amount for all variations at once** from admin:
![Screenshot 2021-03-11 at 16 36 15](https://user-images.githubusercontent.com/2207451/110812482-f060be80-8287-11eb-896c-2d403c485bae.png)

I was a bit surprised to see that the **Low stock report** only checks the stock level vs the store-wide setting (WC > Product > Inventory), so not much to do here really, unless we change the report (which we don't want to do anyway, as it's part of the reports that are being deprecated).

I checked the **Low stock notification email** gets sent correctly when enabled and shows info about the variation:
![Screenshot 2021-03-11 at 16 38 45](https://user-images.githubusercontent.com/2207451/110812889-4a618400-8288-11eb-9940-4020f7921ee3.png)

Additionally, I checked the **CSV Import and Export**, both seem to work fine with this PR.

Finally, I've added the field to the v3 REST API both for GET and POST (to resolve #27371).

Side note: REST API docs need to be regenerated to make sure it includes the new fields.

### How to test the changes in this Pull Request:

1. Create a variation product (P1) with at least 3 variations, set 2 variations to manage stock, one not to. Parent product set to NOT manage stock.
2. Create a variation product (P2) with at least 3 variations, set 2 variations to manage stock, one not to. Parent product set to manage stock.
2. Apply branch.
2. Set the P1.V1 low stock amount to X (different to store-wide setting A), leave the P1.V2 low stock amount to no value (so that it takes the default value from store-wide setting). Set the same for P2 variations, while also setting the product P2's low stock amount to value Z.
3. Set the stock above the limits for these products.
4. Submit orders for the different variations and see if you observe expected behavior.
5. Check that the emails are being sent at the correct stock level for each variation.
6. Export both products, import them back as new products and confirm that the low stock amounts are set correctly.
7. Try to set the low stock amount for all variations via the Bulk Actions dropdown.
8. Check the REST API response for v2 GET requests `http://one.wordpress.test/wp-json/wc/v2/products/P1/` and `http://one.wordpress.test/wp-json/wc/v2/products/P1/variations/V1` to check there are no changes. 
9. Check the REST API response for v3 GET requests `http://one.wordpress.test/wp-json/wc/v3/products/P1/` and `http://one.wordpress.test/wp-json/wc/v3/products/P1/variations/V1` and check if the correct responses get returned, including new data.
9. Try to update the product and variation low stock amounts using a POST request with request body such as `{ "low_stock_amount": 7 }`

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enhancement - Add support for the low stock threshold for variations.

peterfabian committed:
Added low stock threshold input to the Admin UI.

peterfabian committed:
Harmonize the id and name with simple product.

peterfabian committed:
Connect the UI to the db to allow updates of the values.

peterfabian committed:
Don't take the low stock amount info from the parent of the variation, but the variation itself.

Potentially bw incompatible.

peterfabian committed:
Enable bulk updates of variations with low stock amount.

peterfabian committed:
Add support for low stock amount to REST API v3. Ref #27371.

peterfabian committed:
Better wording of the tooltip--being more specific about what the amount relates to.

peterfabian commented on includes/wc-stock-functions.php:
```diff
@@ -395,9 +395,6 @@ function wc_release_stock_for_order( $order ) {
 function wc_get_low_stock_amount( WC_Product $product ) {
-	if ( $product->is_type( 'variation' ) ) {
-		$product = wc_get_product( $product->get_parent_id() );
-	}
```
This is potentially bw incompatible. I could create a new function, but I feel like it should not be a big issue. I scanned the marketplace for usage of this function and don't see any potential problems.

peterfabian committed:
Added comment documenting the history of the function.

peterfabian committed:
Fix existing tests.

Konamiman commented:
I still haven't tested it but I have a couple of suggestions about the UI.

1. The full-width "Low stock threshold" textbox seems a bit overkill. What if you reduce the one for "Stock quantity" to half its width and put that new one next to it?
2. Showing the default value in a lighter shade may be a too subtle way to indicate that a default value is in use. I would keep the shade but instead print the text as "default (2)"
3. Related to 2, the help text could include something like "The default value can be set in Settings - Products - Inventory".

Konamiman commented:
Tested and it works, but I have a couple of remarks.

First, the REST API returns a not set low stock amount as an empty string. Shouldn't it return `null` instead?

Second, the product-wide low stock amount (when the main product is set as "manage stock") is useless: vairations will always use the setting value when no explicit value is set. I'm not sure if this is on purpose, but I think it should work like this:

1. Variation has its own explicit value: use it.
2. Otherwise, main product is set to "manage stock" and has an explicit value set: use it, display as "product default (X)"
3. Otherwise, use setting value, display as "default (X)"

peterfabian committed:
Use null rather than empty string to signify unset value.

peterfabian committed:
Improve the description of the low stock threshold tooltip.

peterfabian commented:
Thanks, great feedback.

> The full-width "Low stock threshold" textbox seems a bit overkill. What if you reduce the one for "Stock quantity" to half its width and put that new one next to it?

Perhaps you're right. I guess I wanted to have a mental distinction, but I agree the inputs are rather wide for many of those values.

> Showing the default value in a lighter shade may be a too subtle way to indicate that a default value is in use. I would keep the shade but instead print the text as "default (2)"

I thought about the same, it seems like an accessibility issue as well. I erred on the side of not changing it as it's how it looks also in regular simple products but agreed it's not very apparent.

> Related to 2, the help text could include something like "The default value can be set in Settings - Products - Inventory".

Sounds like a good idea. Added in 2ae5ec5.

> First, the REST API returns a not set low stock amount as an empty string. Shouldn't it return null instead?

Good point, changed in 8bfa97d.

> Second, the product-wide low stock amount (when the main product is set as "manage stock") is useless: variations will always use the setting value when no explicit value is set. I'm not sure if this is on purpose, but I think it should work like this:

> - Variation has its own explicit value: use it.
> - Otherwise, main product is set to "manage stock" and has an explicit value set: use it, display as "product default (X)"
> - Otherwise, use setting value, display as "default (X)", or maybe "shop default (X)"
And of course the help tooltips should then be updated accordingly.

I was slightly unsure about this, as there is clearly a conflict between the parent product and the variation. I _think_ people often turn off managing stock on variable products (ie the parents) as it doesn't make much sense to track stock on the parent (ie you have 5 T-shirts, but of different colors and sizes). That's why I thought it might make sense to leave this actually somewhat less complex, but I'll check with the community members requesting this feature to double check this.

Konamiman commented:
> I think people often turn off managing stock on variable products (ie the parents) as it doesn't make much sense to track stock on the parent (ie you have 5 T-shirts, but of different colors and sizes). That's why I thought it might make sense to leave this actually somewhat less complex, but I'll check with the community members requesting this feature to double check this.

As i see it, the value for the main product combined with the ability to set a different value for each variation is a good way to implement "I want to get notified when the stock of any variation of this product goes below X, but I still want to be able to set a different threshold for a given variation".

peterfabian committed:
Add parent's default for low stock if notthing is set on individual variation.

I.e. use variation low stock amount value, if not available, use parent product's value, if not available, use the store-wide default.

peterfabian commented:
You're probably right, amended in e6cdd74.

peterfabian committed:
Set correct placeholder value for variation--first the parent, then the store-wide default.

peterfabian committed:
Updated the wording of the tooltip.

Konamiman commented:
New suggested values for the tooltips to provide more clarity on how the entire thing works:

For the product-level value in the Inventory tab:

> When product stock reaches this amount you will be notified by email. It is possible to define different values for each variation individually. The shop default value can be set in Settings > Products > Inventory.

For the variation-level value:

>When variation stock reaches this amount you will be notified by email. The default value for all variations can be set in the product Inventory tab. The shop default value can be set in Settings > Products > Inventory.

As for how to convey that we are using a default value, if you don't want to go with the "default (X)" approach, what about adding a tooltip to the `input` itself (with a `title` attribute)? It could be something like _Using the product default_ / _Using the shop default_.

peterfabian committed:
Make the inherited values explicit by stating where it comes from.

peterfabian committed:
Don't echo the strings, return them for later use.

peterfabian commented:
Thanks for the suggestions, updated. I also changed the placeholder text to show where the value that would get used comes from. Do you think that makes sense?

![Screenshot 2021-03-16 at 13 00 17](https://user-images.githubusercontent.com/2207451/111305940-e3611800-8657-11eb-9dca-57db43fbb521.png)

![Screenshot 2021-03-16 at 13 01 34](https://user-images.githubusercontent.com/2207451/111305957-e78d3580-8657-11eb-83ea-02ba9e6f705a.png)

peterfabian committed:
Improved tests, added save points, fixed comments.

Konamiman commented:
Ok, I've tested everything again with the latest changes and it works fine. To me it now works as it should so I'm merging.

Heatherarts commented:
I currently use https://woocommerce.com/products/variation-shared-inventory-for-woocommerce/
Would this work to fix the issue of not being able to set a low stock notification for shared variation inventory items?
